### PR TITLE
Added a `prefix` parameter when creating a new function.

### PR DESCRIPTION
**Pull request by kenfdev**
Signed-off-by: Ken Fukuyama <kenfdev@gmail.com>

## Description

This adds a `--prefix | -p` parameter to the `faas new` command.
You can set the prefix for the function image name.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
issue #338

## How Has This Been Tested?

```
# command
faas new --lang dockerfile prefixed-func --prefix kenfdev

# output yaml
provider:
  name: faas
  gateway: http://localhost:8080

functions:
  prefixed-func:
    lang: dockerfile
    handler: ./prefixed-func
    image: kenfdev/prefixed-func
```

```
# command
faas new --lang dockerfile shortcut-prefixed-func -p 10.0.95.1:5000

# output yaml
provider:
  name: faas
  gateway: http://localhost:8080

functions:
  shortcut-prefixed-func:
    lang: dockerfile
    handler: ./shortcut-prefixed-func
    image: 10.0.95.1:5000/shortcut-prefixed-func
```

```
# command
faas new --lang dockerfile non-prefixed-func

# output yaml
provider:
  name: faas
  gateway: http://localhost:8080

functions:
  non-prefixed-func:
    lang: dockerfile
    handler: ./non-prefixed-func
    image: non-prefixed-func
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: commands/new_function.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ var (
 func init() {
 	newFunctionCmd.Flags().StringVar(&language, "lang", "", "Language or template to use")
 	newFunctionCmd.Flags().StringVarP(&gateway, "gateway", "g", defaultGateway, "Gateway URL to store in YAML stack file")
+	newFunctionCmd.Flags().StringVarP(&imagePrefix, "prefix", "p", "", "Set prefix for the function image")
```
Comment (kenfdev) [inline commands/new_function.go]: Not sure if the description is good enough

Comment (alexellis) [reply commands/new_function.go]: I think it will work

File: commands/new_function_test.go [hunk 2]
```diff
@@ -25,30 +25,43 @@ const ListOptionOutput = `Languages available as templates:
...
 type NewFunctionTest struct {
-	title       string
-	funcName    string
-	funcLang    string
-	expectedMsg string
+	title         string
+	prefix        string
+	funcName      string
+	funcLang      string
+	expectedImage string
```
Comment (kenfdev) [inline commands/new_function_test.go]: added to test the expected image name (previously it was equal to `funcName`)

File: commands/new_function.go [hunk 3]
```diff
@@ -114,6 +115,13 @@ func runNewFunction(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("got unexpected error while updating .gitignore file: %s", err)
 	}
 
+	var imageName string
+	if len(imagePrefix) > 0 {
+		imageName = imagePrefix + "/" + functionName
```
Comment (alexellis) [inline commands/new_function.go]: What about `strings.Trim()` ? for `  ` whitespace

--prefix="alexellis "

Comment (kenfdev) [reply commands/new_function.go]: Thank you!
Will change it to this.

```
var imageName string
imagePrefix = strings.TrimSpace(imagePrefix)
if len(imagePrefix) > 0 {
	imageName = imagePrefix + "/" + functionName
} else {
	imageName = functionName
}
```

File: commands/new_function.go [hunk 4]
```diff
@@ -114,6 +115,14 @@ func runNewFunction(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("got unexpected error while updating .gitignore file: %s", err)
 	}
 
+	var imageName string
+	imagePrefix = strings.TrimSpace(imagePrefix)
```
Comment (kenfdev) [inline commands/new_function.go]: added trimming of the prefix parameter

Comment (alexellis) [reply commands/new_function.go]: This looks fine. It is also OK to in-line it but I won't get picky.

File: commands/new_function_test.go (hunk 2 above)
Comment (alexellis) [reply commands/new_function_test.go]: I'm impressed. Thank you 💯

**Comment by johnmccabe** [comment]
Should prefix be namespace?

**Comment by alexellis** [comment]
I think prefix works - it is technically the prefix for the image and not the function, but we should be able to live with the ambiguity.

**Pull request merged**